### ISSUE TITLE
SharedViewModel - Buggy Async code

### ISSUE THREAD

**Issue by LavishSwarnkar** [Code3six/FaangX#18]
https://github.com/Code3six/FaangX/blob/0b4924d8cfe8832135b4a60f74daa7917b26683f/app/src/main/java/com/example/faangx/presentation/viewmodel/SharedViewModel.kt#L67-L77

Your `checkUsers()` function checks whether the user profile exists on Firestore or not. And based on that, updates the `SharedViewModel.user` object.

You have called `saveLogin()` function right after `checkUsers()`, which saves the user object in the datastore.

Now here is the catch :

- `checkUsers()` is doing a network operation (inside `viewModelScope.launch`) and is a async task even though the function itself is not suspend.
  
- `saveLogin()` does not wait for `checkUsers()` to complete and hence the updated user object is never saved in the datastore.
  
- In other words, before `checkUsers()` updated the user object, `saveLogin()` saves the user object because `checkUsers()` is a network operation and takes time.
  

## Solution

- Make `checkUsers()` function suspend and don't use `addOnSuccessListener`, otherwise it will have no effect. Instead use `await()`.
  
- Call both `checkUsers()` & `saveLogin()` inside `viewModelScope.launch`
  
- Now `saveLogin()` will not proceed until `checkUsers() `finishes its execution.

**Comment by Code3six** [comment]
I will remove saveLogin() and put it inside checkUsers()
